addons/product/models/product_attribute.py

```diff
@@ -196,6 +196,13 @@ def _update_product_template_attribute_values(self):
             if product_template_attribute_values_to_remove:
                 product_template_attribute_values_to_remove.unlink()
 
+        # delete remaining product.template.attribute.value that are not used on any line
+        for product_template in self.mapped('product_tmpl_id'):
+            self.env['product.template.attribute.value'].search([
+                ('product_tmpl_id', '=', product_template.id),
+                ('product_attribute_value_id', 'not in', product_template.attribute_line_ids.mapped('value_ids').ids),
+            ]).unlink()
+
```

Comment:
Can't the unlink be done once in batch ?
From what I know of the behavior in 12, the fix LGTM.

Reply:
The problem for batching it is that we need to look for obsolete values per template, so to batch it would require to fetch all values for all templates and then filter them in python. I have the feeling this is more CPU work, and definitely more complex code.

Also how often do we batch writing the same values on different attribute lines? I think never.

This makes we think however that this code can probably be moved in `write`, and probably even only if `attribute_id` is in values. It seems useless in `create` and all other cases of `write`.

I'll already wait to see the runbot result on this before moving it.